packages/fxa-auth-server/lib/payments/stripe.js

```diff
@@ -856,6 +856,106 @@ class StripeHelper {
       this.webhookSecret
     );
   }
+
+  /**
+   * Accept a string ID or resource object, return a resource object after
+   * retrieving (if necessary)
+   *
+   * @template T
+   * @param {string | T} resource
+   * @param {string} resourceType
+   *
+   * @returns {Promise<T>}
+   */
+  async expandResource(resource, resourceType) {
+    return typeof resource === 'string'
+      ? this.stripe[resourceType].retrieve(resource)
+      : resource;
+  }
+
+  /**
+   * Extract invoice details for billing emails
+   *
+   * @param {Invoice} invoice
+   */
+  async extractInvoiceDetailsForEmail(invoice) {
+    let uid = '',
+      email = '',
+      invoiceNumber = '',
+      invoiceDate,
+      invoiceTotal = 0,
+      cardType = '',
+      lastFour = '',
+      nextInvoiceDate,
+      productId = '',
+      planId = '',
+      planName = '',
+      planEmailIconURL = '',
+      planDownloadURL = '';
+
+    try {
+      if (invoice && typeof invoice === 'object') {
+        // Dig up & expand objects in the invoice that usually come as just IDs
+        const { plan } = invoice.lines.data[0];
+        const [product, customer, charge] = await Promise.all([
+          this.expandResource(plan.product, 'products'),
```

Comment:
We can't locate this in the product cache?

Reply:
Can probably use the allProducts cache here

Reply:
Reworked things to use AbbrevProduct, so we should hit cache nearly all the time here.